### ISSUE TITLE
[BBB] Port to Beaglebone Black

### ISSUE THREAD

**Issue by michael-ring** [intel-iot-devkit/mraa#78]
Hi,

is anybody already working on a port for the Beaglebone Black?

If yes, fine, if no then I would like to (try to) do the port.

**Comment by desertgourd** [comment]
Hi,
I am trying to get TRMh20's RF24 library working with MRAA for the Beaglebone Black BBB.  TRMh20 has it working with RPi and the github branch can be found at https://github.com/TMRh20/RF24/tree/Updates.  An issue has been started "Porting TMRh20 RF24 Libraries to the BeagleBone Black" and the progress can be followed their.  

My current status is that I have the BBB working with the TMRh20 RF24-updates branch before the MRAA changes.  Moving over to  MRAA I have been successful in installing and building MRAA on the BBB using the following steps.

 TMRh20's steps to install mraa master branch on BBB:
1. /mraa# mkdir build
2. /mraa# cd build
3. /mraa# cmake .. -DBUILDSWIGNODE=OFF
4. /mraa# sudo make install
5. /mraa# sudo ldconfig

This builds and installs all the mraa libraries, headers, and includes.

The RF24 libraries were built using /rf24libs/RF24# make install -B RF24_MRAA=1
Test example programs in the /rf24libs/RF24/examples_RPi directory were built with:
/RF24/examples_RPi# make
/RF24/examples_RPi# make install

When I run the example ./gettingstarted I get the following error:
root@beaglebone:/home/rnh/rf24libs/RF24/examples_RPi# ./gettingstarted
*** glibc detected *** ./gettingstarted: realloc(): invalid next size: 0x00019a08 ***
Aborted

Is this related to the following lines in the file /RF24/arch/MRAA/gpio.cpp :
GPIO::~GPIO() {
// Prophet: this should free memory, and unexport pins when RF24 and/or GPIO gets deleted or goes out of scope
this->close(gpio_ce_pin);
this->close(gpio_cs_pin);
}

I do not fully understand what is happening here and would appreciate any input from this forum.  

As a fall back I would like to try and test my MRAA install with the blink-io.c example. Is there anywhere I could find the BBB MRAA source files for the MRAA examples.  I am looking for IO pin assignments and any setup info or documentation for the MRAA on the BBB. 

Thanks

**Comment by michael-ring** [comment]
I have not yet created the pull request for the beaglebone black so there is no support for the beaglebone available at the moment in mraa. The pull request mentioned in this issue was too big so I split up this work in several pieces.

My first priority was to make the support for the Raspberry Pi complete, but Beaglebone is next, I still need to implement access to gpio via /dev/mem, then I will create the pull request.

mraa compiles on any arm device, but it defaults to think it works on the raspberry pi, perhaps it is better to not do this default behaviour and instead display a message that the device support for the device it is running on is not available.

**Comment by desertgourd** [comment]
@michael-ring
Thanks for the reply.  How is the progress on implementing access to gpio via /dev/mem on the BBB coming?  You mentioned in your Dec. 30 post that you had something working.  Do you have any time frame on how long before a pull request?  Do you need any testers?

**Comment by michael-ring** [comment]
Finishing it as we speak ;-) I can test with led's on Sunday evening, right now I can only test with the 4 on-board led's. 

If you want you can check my forked repository in 2-3 hours for a new commit on branch arm_beaglebone, I am very thankful for help on testing the code.

https://github.com/michael-ring/mraa

**Comment by michael-ring** [comment]
Today's work was pushed to the branch, have fun testing

**Comment by desertgourd** [comment]
I have been looking at the code and will start going over the examples.  Hours of FUN

**Comment by desertgourd** [comment]
@michael-ring
I could not find where to start a new issue on your fork so I am posting this here.
I tried to run the /mraa/examplex/c++/Blink-IO.cpp and got the following error.

root@beaglebone:/home/rnh/mraa/build/examples/c++# ./blink-io-cpp
mraa: FATAL error, failed to initialise platform
Provide an int arg if you want to flash on something other than 8
terminate called after throwing an instance of 'std::invalid_argument'
  what():  Invalid GPIO pin specified
Aborted

I tried 21 - 24 from the following table and none worked.  I used the command line ./blink-io-cpp 22. Is that the right format for passing an argument to the program?

root@beaglebone:/sys/kernel/debug/pinctrl/44e10800.pinmux# cat pingroups
registered pin groups:
group: pinmux_userled_pins
pin 21 (44e10854)
pin 22 (44e10858)
pin 23 (44e1085c)
pin 24 (44e10860)
 
I also tried building the TMRh20 libraries with the mraa option and ran the gettingstarted example with gpio pins set with the command RF24 radio(115,117); (chip enable,chip select), which would be gpio 115 Header#9 pin 27 and gpio 117 Header#9 pin 25 . 
 I no longer got the error:
root@beaglebone:/home/rnh/rf24libs/RF24/examples_RPi# ./gettingstarted
*** glibc detected *** ./gettingstarted: realloc(): invalid next size: 0x00019a08 ***
Aborted

now I get the error:
root@beaglebone:/home/rnh/rf24libs/RF24/examples_RPi# ./gettingstarted
mraa: FATAL error, failed to initialise platform
RF24/examples/GettingStarted/
terminate called after throwing an instance of 'std::invalid_argument'
  what():  Invalid GPIO pin specified
Aborted

I think this is progress.

When the code is working what pin#'s would you use for:
gpio 115 Header#9 pin 27  SPI-CE
gpio 117 Header#9 pin 25  SPI-CSN

Do you have a working example for gpio or spi on the BBB.

**Comment by michael-ring** [comment]
mraa allows you to control the pins available on the extension headers P8 and P9. The UserLeds are not visible on those headers so you cannot change those without hacking the code (as I did for testing).

To get your feet wet you should start with the gpio example, build the code and do:

sudo examples/gpio list (this will show you which pins are available an can get used as gpio)
and
sudo examples/gpio set 7 1 (This will turn pin 7 of Connector P8 to 1)
Connect a led to that pin and you will be able to control it.

What puzzles me is that you get the error:
mraa: Fatal, failed to initialize platform, this should not happen.

can you please post the output of "cat /proc/cpuinfo" and tell me which OS/Kernel you are using on your beaglebone?

I am using Kernel 3.8 with the debian image, this is the official kernel, newer kernels will not work as they do not support capes (for spi and i2c) kernel 3.19 will hopefully solve that problem but is not yet officially supported.

sudo cat /var/log/syslog 

should also give you more details what is going on.

**Comment by desertgourd** [comment]
Running the gpio example gave the following output:
root@beaglebone:/home/rnh/mraa/build/examples# ./gpio set 7 1
mraa: FATAL error, failed to initialise platform
Could not initialize gpio 7

System Info:
cat /proc/cpuinfo
processor       : 0
model name      : ARMv7 Processor rev 2 (v7l)
BogoMIPS        : 298.24
Features        : swp half thumb fastmult vfp edsp thumbee neon vfpv3 tls
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x3
CPU part        : 0xc08
CPU revision    : 2

Hardware        : Generic AM33XX (Flattened Device Tree)
Revision        : 0000
Serial          : 0000000000000000
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
login as: root
Debian GNU/Linux 7

BeagleBoard.org BeagleBone Debian Image 2014-05-14

Support/FAQ: http://elinux.org/Beagleboard:BeagleBoneBlack_Debian
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
BeagleBone Black Rev. C with latest official Debian image 2014/5/14

Is there a command that would list OS Kernel version info?

**Comment by desertgourd** [comment]
Output from cat /var/log/syslog  after running  ./gpio set 7 1
Feb 21 17:17:02 beaglebone /USR/SBIN/CRON[9707]: (root) CMD (   cd / && run-parts --report /etc/cron.hourly)
Feb 21 17:34:27 beaglebone libmraa[9920]: libmraa version HEAD-HASH-NOTFOUND initialised by user 'root' with EUID 0
Feb 21 17:34:27 beaglebone libmraa[9920]: Beaglebone: failed to initialize
Feb 21 17:34:27 beaglebone libmraa[9920]: gpio: platform not initialised

**Comment by desertgourd** [comment]
If 'examples/gpio set 7 1'  turns pin 7 of Connector P8 to 1 then how do you address pin 7 of connector P9

**Comment by michael-ring** [comment]
Kernel version: uname -a
P9: Each of the connectors has 46 pins, you will need to add 46 to the pin number to reach P9 --> Pin1 P9 = 47 in mraa. 

I am a little puzzled about the failed to initialize message, this is the root cause of your problems.

I have a Beaglebone Rev C at home, but not with me right now so I cannot check.

Please follow the steps listed below that we make 100% sure we have the same software revision and build configuration:

create an empty directory: mkdir xxxxx
change there: cd xxxxx
git clone -b arm_beaglebone https://github.com/michael-ring/mraa.git
cd mraa
mkdir build
cd build
cmake -DCMAKE_BUILD_TYPE=Debug ..
make
sudo examples/gpio version

should give you:
Version v0.5.4-142-gb46146a on Beaglebone Black Rev. C

if not something is wrong with initialization.

if you feel you cannot wait until this evening it would be helpfull if you could debug a little.

I found it the easiest to use cgdb for debugging:
sudo apt-get install gdb cgdb

now go to the build directory and write:
sudo cgdb --args examples/gpio version

You are now in cgdb, write:
(gdb) break beaglebone.c:mraa_beaglebone
No source file named beaglebone.c.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (beaglebone.c:mraa_beaglebone) pending.
(gdb) run

gdb will now break on init of beaglebone.

you can hit esc-key then F8 and F10 can be used to single step / step over

follow the code to see why the error line is reached.

If you want to inspect vars hit 'i'. hitting ESC brings you back to the mode where F8 and F10 work

**Comment by desertgourd** [comment]
Thank you for the help. I followed your instructions on debuging (gdb and cgdb really great, learning new stuff all the time) and found the following error.

mraa/src/arm/beaglebone.c
```c++
460   if (is_rev_c == 1) {
461       //b->platform_name, PLATFORM_NAME_BEAGLEBONE_BLACK_REV_C;  had , instead of =
462       b->platform_name = PLATFORM_NAME_BEAGLEBONE_BLACK_REV_C; // changed , to =
463       b->phy_pin_count = MRAA_BEAGLEBONE_BLACK_PINCOUNT;
464   }
```
it appears to be working, when I run examples.gpio list  it list out the pins and their assignments.  I will continue to check it out.
Thanks again

**Comment by desertgourd** [comment]
I have tested the gpio pin outputs with a logic probe and I can use 'gpio set pin value' to high or low.  

What is the status of the SPI interface, I tried the example/spi_max7219 but get:
Initialization of spi failed, check syslog for details, exit...
I do not have a max7219 connected but wanted to see if I was getting anything on the CE and CSN pins with the logic probe.

syslog:
Feb 22 16:24:39 beaglebone libmraa[29464]: libmraa version v0.5.4-142-gb46146a initialised by user 'root' with EUID 0
Feb 22 16:24:39 beaglebone kernel: [521295.041555] bone-capemgr bone_capemgr.9: part_number 'BB-SPI0-01', version 'N/A'
Feb 22 16:24:39 beaglebone kernel: [521295.041737] bone-capemgr bone_capemgr.9: slot #10: generic override
Feb 22 16:24:39 beaglebone kernel: [521295.041784] bone-capemgr bone_capemgr.9: bone: Using override eeprom data at slot 10
Feb 22 16:24:39 beaglebone kernel: [521295.041833] bone-capemgr bone_capemgr.9: slot #10: 'Override Board Name,00A0,Override Manuf,BB-SPI0-01'
Feb 22 16:24:39 beaglebone kernel: [521295.042088] bone-capemgr bone_capemgr.9: slot #10: Requesting part number/version based 'BB-SPI0-01-00A0.dtbo
Feb 22 16:24:39 beaglebone kernel: [521295.042137] bone-capemgr bone_capemgr.9: slot #10: Requesting firmware 'BB-SPI0-01-00A0.dtbo' for board-name 'Override Board Name', version '00A0'
Feb 22 16:24:40 beaglebone kernel: [521295.077531] bone-capemgr bone_capemgr.9: failed to load firmware 'BB-SPI0-01-00A0.dtbo'
Feb 22 16:24:40 beaglebone libmraa[29464]: spi: Device not initialized

**Comment by michael-ring** [comment]
For SPI:

Have a look at 

http://elinux.org/BeagleBone_Black_Enable_SPIDEV

You will need BB-SPI0-01-00A0.dtbo and BB-SPI1-01-00A0.dtbo in /lib/firmware to be able to use SPI, the overlay first needs to get loaded before SPI is available.

If you have trouble creating those files yourself you can download them here:

http://temp.michael-ring.org/BB-SPI0-01-00A0.dtbo
http://temp.michael-ring.org/BB-SPI1-01-00A0.dtbo

**Comment by desertgourd** [comment]
Hi, I added the .dtbo files and the spi initializes when I start the example/spi_max7219.  

When I try to use them with the TRMh20 RF24 libraries the spi initializes but it is not reading/writing data to the RF24 spi module.  The logic probe shows activity on the CSN and MISO pins but not on any others.  I believe this is caused by the .dtbo files not configured for the right pins.  I am working on configuring the .dts and building the .dtbo for the TRMh20 hardware configuration.

**Comment by michael-ring** [comment]
Which port are you using, spi0 can only be used when hdmi is disabled...

Von meinem iPhone gesendet

**Comment by desertgourd** [comment]
I have the RF24 board installed on SPI1 and have disabled the HDMI. It was working on that port with CSN on P9-25 and CE on P9-27 with the TRMh20 RF24 library. 

Your arm_beaglebone branch is using CSN P9-28, I have checked the documentation and that is correct for SPI1.  I will change my wiring to conform to that.  

On the CE pin I am using P9-27 which is initialized as:
mraa/scr/arm/beaglebone.c
1163    strncpy(b->pins[73].name, "GPIO125", 8); 
1164    b->pins[73].capabilites = (mraa_pincapabilities_t) {1, 1, 0, 0, 0, 0, 0, 0};
1165    b->pins[73].gpio.pinmap = 125;  

I have found P9-27 to be listed as either GPIO125 pinmap=125 or GPIO115 pinmap=115 depending on which reference you were looking at. The best reference I have found for this is https://github.com/derekmolloy/boneDeviceTree/tree/master/docs check out the P9 header table.  He lists the P9-27 pin as GPIO115 and this is what we found to work with the TMRh20 RF24 library. I will check out both configurations but have not got to that yet.

When I run the gettingstarted example of the TMRh20 RF24 library the spi should be working but I get no logic probe activity on SCLK pin 77 or D1 MOSI pin 76, I do get activity on pin 71 and 75.

When I run the example gpio set pin value on pins 74, 75, 76, 77 I get 'Could not initialize gpio '.  It dose work on pins 7, 71, 73.

Dose anything have to be done besides copying file BB-SPI1-01-00A0.dtbo to /lib/firmware.  Dose MRAA load the overlays when the spi and gpio examples are run, and dose it disable the overlays when the program closes.

**Comment by michael-ring** [comment]
I checked SPI Interface tonight with my logic analyzer. Here are the Results:

SPI0 works
CS = P9.17
D1 = P9.18 This is Data Out, has to be connected to the Input of your device
D0 = P9.21 This is Data In, has to be connected with the output of your device
SCLK = P9.22 The Clock Signal

SPI1 works
CS=P9.28 is CS
D1 = P9.30 Data Out
D0 = P9.29 Data In
SCLK = P9.31

Both results based on the dts I put on upload. 

Here's a picture that shows the layout:
http://elinux.org/images/b/bf/2_SPI_Ports.PNG

In the picture it also says that P9-27 is 125, but you are right, 115 is also correct for me, thank you for pointing this out, I have fixed my code.

When emmc is disabled then the pins 74..77 can get used as either GPIO or SPI. once you use them as SPI when loading the kernel module you cannot use them as GPIO anymore. The pin function is only evaluated on startup of mraa, so both options are visible. Try loading SPI device manually before you start mraa, then you will see in gpio list that the gpio function for those pins is not available anymore.

Overlays are loaded by mraa but not unloaded, especially unloading the SPI overlay for example can crash your beaglebone for Kernels <3.19. For i2c and usart this is doable, but from my point of view this should wait until spi unloading also works fine, this way the behaviour is at least consistent. 

Sorry to come back to you this late, my real jobs is keeping me quite busy this week....

**Comment by desertgourd** [comment]
Where are the .dts files of  BB-SPI0-01-00A0.dtbo and BB-SPI1-01-00A0.dtbo so I can see how they are structured.  I have been debugging today and found that the SPI.dtbo files from running the TMRh20 libraries were still active and had to be disabled before trying the MRAA libraries.  I will reboot the BBB and copy  your  BB-SPI0-01-00A0.dtbo and BB-SPI1-01-00A0.dtbo files to /lib/firmware and test the MRAA.
Thanks for the help.

**Comment by michael-ring** [comment]
You can download them from http://elinux.org/BeagleBone_Black_Enable_SPIDEV , they then have .txt ending, but those are the dts files. Make sure to grab the right versions that define D1 as output

**Comment by desertgourd** [comment]
I am starting to get some positive response from the SPI interface to the nRF24L01 board.  I had both BB-SPI0-01-00A0.dtbo and BB-SPI1-01-00A0.dtbo installed to /lib/firmware, with the RF24 board on the SPI1 pins.  The MRAA is only initializeing SPI0.  SPI1 never gets activated.  I moved the RF24 board to the SPI0 pins and am getting pulse activity on the CSN, SCLK, MOSI and MISO pins with the logic probe.  I need to connect the CE pin for output from the BBB to the RF24 board.  I am wondering what pin would be a good choise and have the least amount of conflicts with other configerations.

**Comment by desertgourd** [comment]
Is there some way to specifiy the SPI1 to be used instead of SPI0.

**Comment by prophet-roshak** [comment]
Hi!
SPI0 is currently hardcoded in rf24 lib, you can change it in
arch/mraa/spi.c
25 февр. 2015 г. 4:00 пользователь "desertgourd" <notifications@github.com>
написал:

**Comment by desertgourd** [comment]
@michael-ring  and  @TMRh20

I have the MRAA working with the TMRh20 RF24 libraries and have tested the 'gettingstarted' and 'transfer' examples in the /examples_RPi directory.  In the process I moved the RF24 board to the default SPI0 pins:
P9-17___CSN 
P9-18___D1_MOSI
P9-21___D0_MISO
P9-22___SCLKP
The CE pin was assigned to:
P9-13___CE

The TRMh20 RF24 constructor sets up the CE and CSN with the command 'RF24 radio(CE-pin, CSN-pin);' with the pin numbering format  as follows:
pin#=P8 Header Pin
pin#=P9 Header Pin + 46  
So the RF24 constructor would be 'RF24 radio(59,63);'.  The problem is that this dose not work because pin 63 is not a valid GPIO pin at this stage. Trying to run the transfer example generates the error:

root@beaglebone:/home/rnh/rf24libs/RF24/examples_RPi# ./transfer
RF24/examples/Transfer/
terminate called after throwing an instance of 'std::invalid_argument'
  what():  Invalid GPIO pin specified
Aborted

gdb debuger output:
(gdb) r
Starting program: /home/rnh/rf24libs/RF24/examples_RPi/transfer
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/arm-linux-gnueabihf/libthread_db.so.1".
RF24/examples/Transfer/
terminate called after throwing an instance of 'std::invalid_argument'
  what():  Invalid GPIO pin specified

Program received signal SIGABRT, Aborted.
0xb6db9f96 in ?? () from /lib/arm-linux-gnueabihf/libc.so.6
(gdb) Quit
(gdb)

syslog output:
Feb 26 16:39:30 beaglebone libmraa[2402]: libmraa version v0.5.4-142-gb46146a initialised by user 'root' with EUID 0
Feb 26 16:39:30 beaglebone libmraa[2402]: gpio: pin 63 not capable of gpio

Even when the BB-SPI0-01-00A0.dtbo overlay file has been enabled to setup the SPIO pins I still get the error. I got around the problem by assigning the CSN-pin to P8-8 which is a valid GPIO pin at boot with 'RF24 radio(59,8);'.  The SPIO still outputs the CSN signal on P9-17 that was setup by the BB-SPI0-01-00A0.dtbo overlay file and where the CSN wire for the RF24 board is connected.  

For the TMRh20 with the MRAA framework the CSN pin in the 'RF24 radio(59,8);' statement is redundent and generates errors.  Is there any way to change the MRAA program so only the CE pin is passed 'RF24 radio(59);'?  This fix will also require a change in the TMRh20 RF24 library.

The same problem was occurring when I had the CE pin assigned to P9-15 that was an invalid GPIO pin at boot and I moved it to P9-13.

**Comment by desertgourd** [comment]
@prophet-roshak
Thanks for the information on the where the SPI0 is defined.  I found it in the arch/mraa/spi.c and find it useful.

**Comment by michael-ring** [comment]
SPI as it is currently implemented uses the default mechanism provided by the linux kernel driver which means that CS is managed by the kernel completely on the pin mentioned in one of my last posts.

You can perhaps define a gpio pin for TRM20h to make the API happy, but you should not try to use the pin that is associated to the SPI Port as this pin is not available to you when you use mraa because the kernel driver owns it.

I am however thinking about doing some tests to tell the kernel to not manage CS, I need this for the Raspberry Pi, but if this works it should also work on the Beaglebone. If I succeed you will be able to use any free GPIO Pin as CS and mraa will handle the rest if you are able to use the calls provided by mraa. 
Weather forecast is bad for Sunday so it might be that I work on that, but no promise.

**Comment by desertgourd** [comment]
I did some testing to check performance of the TMRh20 RF24 library with the MRAA framework. Their were 3 variables; the spi-max-frequency set in the device tree overlay file with a default of 24 MHz, The mspi frequency set in the /RF24/arch/mraa/spi.cpp, and the Air Data Rate set with radio.setDataRate(RF24_2MBPS); in the gettingstarted and transfer example files the /RF24/examples_RPi directory.

BB-SPI0-01-00A0.dtbo device tree overlay spi-max-frequency = 24000000
/arch/mraa/spi.cpp mspi->frequency(8000000);
radio.setDataRate(RF24_2MBPS);
gettingstarted__Nano->BBB 6ms delay, no fails
gettingstarted__BBB->Nano 7ms delay, no fails
transfer________Nano->BBB 36KB/S no drops steady rate
transfer________BBB->Nano 36KB/S no drops 10K packets, 1K packets 22-50KB/S
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
BB-SPI0-01-00A0.dtbo device tree overlay spi-max-frequency = 24000000
/arch/mraa/spi.cpp mspi->frequency(8000000);
radio.setDataRate(RF24_1MBPS);
gettingstarted__Nano->BBB 6ms delay, no fails
gettingstarted__BBB->Nano 7ms delay, no fails
transfer________Nano->BBB 25KB/S no drops steady rate
transfer________BBB->Nano 5KB/S 600 dropped packets out of 10K packets
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
BB-SPI0-01-00A0.dtbo device tree overlay spi-max-frequency = 24000000
radio.setDataRate(RF24_2MBPS);
gettingstarted__Nano->BBB 6ms delay, no fails
gettingstarted__BBB->Nano 7ms delay, no fails
transfer________Nano->BBB 37KB/S no drops steady rate
transfer________BBB->Nano 39KB/S no drops 10K packets, 1K packets 34-54KB/S
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
BB-SPI0-01-00A0.dtbo device tree overlay spi-max-frequency = 24000000
radio.setDataRate(RF24_1MBPS);
gettingstarted__Nano->BBB 6ms delay, no fails
gettingstarted__BBB->Nano 8ms delay, no fails
transfer________Nano->BBB 25KB/S no drops steady rate
transfer________BBB->Nano 5KB/S 400 dropped packets out of 10K packets
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
BB-SPI0-01-00A0.dtbo device tree overlay spi-max-frequency = 8000000
/arch/mraa/spi.cpp mspi->frequency(8000000);
radio.setDataRate(RF24_2MBPS);
gettingstarted__Nano->BBB 6ms delay, no fails
gettingstarted__BBB->Nano 7ms delay, no fails
transfer________Nano->BBB 36KB/S no drops steady rate
transfer________BBB->Nano 32KB/S no drops 10K packets, 1K packets 28-52KB/S
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
BB-SPI0-01-00A0.dtbo device tree overlay spi-max-frequency = 8000000
/arch/mraa/spi.cpp mspi->frequency(16000000);
radio.setDataRate(RF24_1MBPS);
gettingstarted__Nano->BBB 6ms delay, no fails
gettingstarted__BBB->Nano 8ms delay, no fails
transfer________Nano->BBB 25KB/S no drops steady rate
transfer________BBB->Nano 5KB/S 400 dropped packets out of 10K packets
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

Conclusions:
1. The device tree overlay spi-max-frequency has no effect and can be left at the 24 MHz default.
2. The mspi frequency set in the /arch/mraa/spi.cpp file has no effect and can be left a 8 MHz.
3 The Air data rate has a grate effect with 1MBPS having lower transfer rates and considerable dropout. The 2MBPS gives an average 10K packet rate of 36 KB/S and the 1K packet rate at the receiver varied from 22-54 KB/S. Their are probably timing issues with the 1MBPS rate. The 2MBPS may have some timing issues preventing a steady 54 KB/S transfer rate.

**Comment by desertgourd** [comment]
@michael-ring 
Thanks for the help on getting the MRAA working with the TMRh20 RF24 library on the BBB.  Based on suggestion from you TRMh20 made changes and everything is working great.

MRAA, 1MBPS, Transfer = 47.8 KBPS, No Drops / 10K Packets, Rock Solid Connection
MRAA, 2MBPS, Transfer = 68 KBPS, No Drops / 10K Packets, Rock Solid Connection

**Comment by tingleby** [comment]
As far as I'm aware the BBB code is all good. I'm going to go ahead and close this. Comment if you want it reopened.
